File: UnitTestWinRTCompCpp/unittestWinRTComp.cpp
#include "pch.h"
#include <Windows.Foundation.h>
#include <wrl\wrappers\corewrappers.h>
#include <wrl\client.h>
#include <stdio.h>
#include "..\WRLCompV1\WRLCompV1_h.h"
#include "CppUnitTest.h"

using namespace Microsoft::VisualStudio::CppUnitTestFramework;
using namespace ABI::Windows::Foundation;
using namespace Microsoft::WRL;
using namespace Microsoft::WRL::Wrappers;

#define TEXT_STR1		L"Owen Farm"
#define TEXT_ZIP1		L"04400"
#define TEXT_CTY1		L"Mos Eisly"
#define TEXT_STR2		L"Senate Apartment Complex"
#define TEXT_ZIP2		L"83100"
#define TEXT_CTY2		L"Coruscant"
#define TEXT_NAME1		L"Luke"
#define TEXT_SURNAME1	L"Skywlaker"
#define TEXT_NAME2		L"Darth"
#define TEXT_SURNAME2	L"Vador"

namespace UnitTestWinRTCompCpp
{
    TEST_CLASS(UnitTestWinRTComp)
    {
    public:
        TEST_METHOD(TestMethodCitizenV2)
        {
			WinRTCompV2::ICitizen^ citizenV2 = ref new WinRTCompV2::Citizen(TEXT_NAME1, TEXT_SURNAME1);
			WinRTCompV2::IAddress^ addressV2 = ref new WinRTCompV2::Address();

			citizenV2->Address = addressV2;
			addressV2->City = TEXT_CTY1;
			Assert::AreEqual(TEXT_CTY1, citizenV2->Address->City->Data());

			WinRTCompV2::IAddress^ addressV3 = ref new WinRTCompV3::Address();

			citizenV2->Address = addressV3;
			addressV3->City = TEXT_CTY2;
			Assert::AreNotEqual(TEXT_CTY2, citizenV2->Address->City->Data());
        }

        TEST_METHOD(TestMethodCitizenV2WithAddressV2b)
        {
			WinRTCompV2::ICitizen^ citizenV2 = ref new WinRTCompV2::Citizen(TEXT_NAME1, TEXT_SURNAME1);
			WinRTCompV2::IAddress^ addressV2 = ref new WinRTCompV2::Address();

			citizenV2->Address = addressV2;
			addressV2->City = TEXT_CTY1;
			Assert::AreEqual(TEXT_CTY1, citizenV2->Address->City->Data());

			WinRTCompV2::IAddress^ addressV2b = ref new WinRTCompV2::Address2();

			citizenV2->Address = addressV2b;
			addressV2b->City = TEXT_CTY2;
			Assert::AreNotEqual(TEXT_CTY2, citizenV2->Address->City->Data());
        }

		/**
			Test of the WRL component Person that implements IPerson and ISaveable

			This test uses the C++/CX extensions create and use the Person component
		 */
		TEST_METHOD(TestPersonWRLCompV1)
		{
			WRLCompV1::PersonClass^ personWrl = ref new WRLCompV1::PersonClass();

			personWrl->Name = TEXT_NAME1;
			personWrl->Surname = TEXT_SURNAME1;

			Assert::AreEqual(TEXT_NAME1, personWrl->Name);
			Assert::AreEqual(TEXT_SURNAME1, personWrl->Surname);

			WRLCompV1::ISaveable^ saveable = personWrl; 
			Assert::IsTrue(saveable->CanSave());
		}

		/**
			This test directly uses the WRL framework to create and use the Person component
		 */
		TEST_METHOD(TestCOMLikeLoadOfWinRTComponent)
		{
			// Activate an instance of Person and get the IPerson interface
			HStringReference runtimeClassPerson(L"WRLCompV1.PersonClass");
			ComPtr<ABI::WRLCompV1::IPerson> pIPersonWrl;
			HRESULT hr = Windows::Foundation::ActivateInstance< ComPtr<ABI::WRLCompV1::IPerson>>( runtimeClassPerson.Get(), &pIPersonWrl );
			Assert::IsTrue(SUCCEEDED(hr));

			HString name;
			HString surname;
			name.Set(TEXT_NAME1);
			surname.Set(TEXT_SURNAME1);

			// Set the name and the surname
			hr = pIPersonWrl->put_Name(name.Get());
			Assert::IsTrue(SUCCEEDED(hr));				

			hr = pIPersonWrl->put_Surname(surname.Get());
			Assert::IsTrue(SUCCEEDED(hr));		

			HSTRING hName;
			HSTRING hSurname;

			// Get the name and the surname
			hr = pIPersonWrl->get_Name(&hName);
			Assert::IsTrue(SUCCEEDED(hr));		

			hr = pIPersonWrl->get_Surname(&hSurname);
			Assert::IsTrue(SUCCEEDED(hr));

			name.Set(hName);
			surname.Set(hSurname);

			// Validate the test
			Assert::AreEqual(TEXT_NAME1, name.GetRawBuffer(nullptr));
			Assert::AreEqual(TEXT_SURNAME1, surname.GetRawBuffer(nullptr));

			// Get the ISaveable interface
			ComPtr<ABI::WRLCompV1::ISaveable> pISaveableWrl;
			hr = pIPersonWrl.As(&pISaveableWrl);

			Assert::IsTrue(SUCCEEDED(hr));

			// Call the can save method
			boolean canSave;
			hr = pISaveableWrl->CanSave(&canSave);

			Assert::IsTrue(canSave);
		}
    };
}

File: UnitTestWinRTComp/UnitTestWinRTComp.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Microsoft.VisualStudio.TestPlatform.UnitTestFramework;

namespace UnitTestWinRTComp
{
    static class TestData
    {
        public const string
            TEXT_STR1 = "Owen Farm",
            TEXT_ZIP1 = "04400",
            TEXT_CTY1 = "Mos Eisly",
            TEXT_STR2 = "Senate Apartment Complex",
            TEXT_ZIP2 = "83100",
            TEXT_CTY2 = "Coruscant",
            TEXT_NAME1 = "Luke",
            TEXT_SURNAME1 = "Skywlaker",
            TEXT_NAME2 = "Darth",
            TEXT_SURNAME2 = "Vador";
    }

    [TestClass]
    public class UnitTestMixCompV2V2b
    {

        [TestMethod]
        public void UnitTestMixingCompV2withCompV2b()
        {
            WinRTCompV2.IAddress addressV2 = new WinRTCompV2.Address(TestData.TEXT_STR1, TestData.TEXT_ZIP1, TestData.TEXT_CTY1);
            WinRTCompV2.IAddress addressV2b = new WinRTCompV2.Address2(TestData.TEXT_STR1, TestData.TEXT_ZIP1, TestData.TEXT_CTY1);

            WinRTCompV2.ICitizen citizenV2 = new WinRTCompV2.Citizen(TestData.TEXT_NAME1, TestData.TEXT_SURNAME1);
            WinRTCompV2.ICitizen citizenV2b = new WinRTCompV2.Citizen2(TestData.TEXT_NAME2, TestData.TEXT_SURNAME2);

            citizenV2.Address = addressV2b;
            addressV2b.ZipCode = TestData.TEXT_ZIP2;
            string zip2 = citizenV2b.Address.ZipCode;
            Assert.AreNotEqual(TestData.TEXT_ZIP2, zip2);

            citizenV2b.Address = addressV2;
            addressV2b.ZipCode = TestData.TEXT_ZIP1;
            string zip2b = citizenV2.Address.ZipCode;
            Assert.AreEqual(TestData.TEXT_ZIP1, zip2b);
        }
    }

    [TestClass]
    public class UnitTestWinRTCompMixV2V3
    {
        [TestMethod]
        public void UnitTestMixingWinRTCompV2withWinRTCompV3()
        {
            WinRTCompV2.IAddress addressV2 = new WinRTCompV2.Address(TestData.TEXT_STR1, TestData.TEXT_ZIP1, TestData.TEXT_CTY1);
            WinRTCompV2.IAddress addressV3 = new WinRTCompV3.Address(TestData.TEXT_STR2, TestData.TEXT_ZIP2, TestData.TEXT_CTY2);

            WinRTCompV2.ICitizen citizenV2 = new WinRTCompV2.Citizen(TestData.TEXT_NAME1, TestData.TEXT_SURNAME1);
            WinRTCompV2.ICitizen citizenV3 = new WinRTCompV3.Citizen(TestData.TEXT_NAME2, TestData.TEXT_SURNAME2);

            citizenV2.Address = addressV2;
            addressV2.ZipCode = TestData.TEXT_ZIP2;

            string zip2 = citizenV2.Address.ZipCode;
            Assert.AreEqual(TestData.TEXT_ZIP2, zip2);

            citizenV3.Address = addressV3;
            addressV3.ZipCode = TestData.TEXT_ZIP1;

            string zip3 = citizenV3.Address.ZipCode;
            Assert.AreEqual(TestData.TEXT_ZIP1, zip3);

            // Mixing Address V3 with Citizen V2
            citizenV2.Address = addressV3;
            addressV3.Street = TestData.TEXT_STR1;

            // The Street property has been changed AFTER the Address V3 was used for the Citizen V2
            string street2 = citizenV2.Address.Street;

            // Because it was not possible to create the association at the composition classes level,
            // the Street value in Citizen.Address is not updated.
            Assert.IsFalse(TestData.TEXT_STR1 == citizenV2.Address.Street);

            // However it is possible to mix V2 with V3 because in the V3 implementation the association
            // is between WinRT instances only
            citizenV3.Address = addressV2;
            addressV2.City = TestData.TEXT_CTY2;
            string city3 = citizenV3.Address.City;
            Assert.IsTrue(TestData.TEXT_CTY2 == citizenV3.Address.City);
        }
    }

    [TestClass]
    public class UnitTestWinRTCompV3
    {
        [TestMethod]
        public void TestPersonV3()
        {
            WinRTCompV2.IPerson person = new WinRTCompV3.Person();

            Assert.IsFalse(person.CanSave());

            person.Name = TestData.TEXT_NAME1;
            person.Surname = TestData.TEXT_SURNAME1;

            Assert.IsTrue(person.CanSave());

            Assert.AreEqual(TestData.TEXT_NAME1, person.Name);
            Assert.AreEqual(TestData.TEXT_SURNAME1, person.Surname);

            person = new WinRTCompV3.Person(TestData.TEXT_NAME2, TestData.TEXT_SURNAME1);

            Assert.IsTrue(person.CanSave());

            Assert.AreEqual(TestData.TEXT_NAME2, person.Name);
            Assert.AreEqual(TestData.TEXT_SURNAME1, person.Surname);

            person.Name = TestData.TEXT_NAME1;

            Assert.AreEqual(TestData.TEXT_NAME1, person.Name);
            Assert.AreEqual(TestData.TEXT_SURNAME1, person.Surname);
        }

        [TestMethod]
        public void TestAddressV3()
        {
            WinRTCompV2.IAddress address = new WinRTCompV3.Address();
            Assert.AreEqual("", address.Street);

            Assert.IsFalse(address.CanSave());

            address.Street = TestData.TEXT_STR1;
            address.ZipCode = TestData.TEXT_ZIP1;
            address.City = TestData.TEXT_CTY1;

            Assert.AreEqual(TestData.TEXT_STR1, address.Street);
            Assert.AreEqual(TestData.TEXT_ZIP1, address.ZipCode);
            Assert.AreEqual(TestData.TEXT_CTY1, address.City);

            address.Street = TestData.TEXT_STR2;
            address.ZipCode = TestData.TEXT_ZIP2;
            address.City = TestData.TEXT_CTY2;

            Assert.AreEqual(TestData.TEXT_STR2, address.Street);
            Assert.AreEqual(TestData.TEXT_ZIP2, address.ZipCode);
            Assert.AreEqual(TestData.TEXT_CTY2, address.City);

            Assert.IsTrue(address.CanSave());
        }

        [TestMethod]
        public void TestCitizenV3()
        {
            WinRTCompV2.ICitizen citizen1 = new WinRTCompV3.Citizen(TestData.TEXT_NAME1, TestData.TEXT_SURNAME1);

            Assert.IsFalse(citizen1.CanSave());
            Assert.AreEqual(TestData.TEXT_NAME1, citizen1.Name);
            Assert.AreEqual(TestData.TEXT_SURNAME1, citizen1.Surname);

            WinRTCompV2.IAddress address = citizen1.Address;
            address.City = TestData.TEXT_CTY1;
            address.Street = TestData.TEXT_STR1;
            address.ZipCode = TestData.TEXT_ZIP1;

            Assert.IsTrue(citizen1.CanSave());

            Assert.AreEqual(TestData.TEXT_CTY1, citizen1.Address.City);
            Assert.AreEqual(TestData.TEXT_STR1, citizen1.Address.Street);
            Assert.AreEqual(TestData.TEXT_ZIP1, citizen1.Address.ZipCode);

            WinRTCompV3.Address address2 = new WinRTCompV3.Address();
            WinRTCompV2.ICitizen citizen2 = new WinRTCompV3.Citizen(TestData.TEXT_NAME2, TestData.TEXT_SURNAME2, address2);

            Assert.IsFalse(citizen2.CanSave());
            Assert.AreEqual(TestData.TEXT_NAME2, citizen2.Name);
            Assert.AreEqual(TestData.TEXT_SURNAME2, citizen2.Surname);

            address2.City = TestData.TEXT_CTY2;
            address2.Street = TestData.TEXT_STR2;
            address2.ZipCode = TestData.TEXT_ZIP2;

            Assert.AreEqual(TestData.TEXT_CTY2, citizen2.Address.City);
            Assert.AreEqual(TestData.TEXT_STR2, citizen2.Address.Street);
            Assert.AreEqual(TestData.TEXT_ZIP2, citizen2.Address.ZipCode);
        }
    }

    [TestClass]
    public class UnitTestWRLCompV1
    {
        [TestMethod]
        public void TestWRLPerson()
        {
            WRLCompV1.PersonClass person = new WRLCompV1.PersonClass();

            person.Name = TestData.TEXT_NAME1;
            person.Surname = TestData.TEXT_SURNAME1;

            Assert.AreEqual(TestData.TEXT_NAME1, person.Name);
            Assert.AreEqual(TestData.TEXT_SURNAME1, person.Surname);
        }
    }

    [TestClass]
    public class UnitTestWinRTCompV2
    {
        [TestMethod]
        public void TestAddressV2()
        {
            WinRTCompV2.IAddress address = new WinRTCompV2.Address();
            Assert.AreEqual("", address.Street);

            Assert.IsFalse(address.CanSave());

            address.Street = TestData.TEXT_STR1;
            address.ZipCode = TestData.TEXT_ZIP1;
            address.City = TestData.TEXT_CTY1;

            Assert.AreEqual(TestData.TEXT_STR1, address.Street);
            Assert.AreEqual(TestData.TEXT_ZIP1, address.ZipCode);
            Assert.AreEqual(TestData.TEXT_CTY1, address.City);

            address.Street = TestData.TEXT_STR2;
            address.ZipCode = TestData.TEXT_ZIP2;
            address.City = TestData.TEXT_CTY2;

            Assert.AreEqual(TestData.TEXT_STR2, address.Street);
            Assert.AreEqual(TestData.TEXT_ZIP2, address.ZipCode);
            Assert.AreEqual(TestData.TEXT_CTY2, address.City);

            Assert.IsTrue(address.CanSave());
        }

        [TestMethod]
        public void TestPersonV2()
        {
            WinRTCompV2.IPerson person = new WinRTCompV2.Person();

            Assert.IsFalse(person.CanSave());

            person.Name = TestData.TEXT_NAME1;
            person.Surname = TestData.TEXT_SURNAME1;

            Assert.IsTrue(person.CanSave());

            Assert.AreEqual(TestData.TEXT_NAME1, person.Name);
            Assert.AreEqual(TestData.TEXT_SURNAME1, person.Surname);

            person = new WinRTCompV2.Person(TestData.TEXT_NAME2, TestData.TEXT_SURNAME1);

            Assert.IsTrue(person.CanSave());

            Assert.AreEqual(TestData.TEXT_NAME2, person.Name);
            Assert.AreEqual(TestData.TEXT_SURNAME1, person.Surname);

            person.Name = TestData.TEXT_NAME1;

            Assert.AreEqual(TestData.TEXT_NAME1, person.Name);
            Assert.AreEqual(TestData.TEXT_SURNAME1, person.Surname);
        }

        [TestMethod]
        public void TestCitizenV2()
        {
            WinRTCompV2.ICitizen citizen1 = new WinRTCompV2.Citizen(TestData.TEXT_NAME1, TestData.TEXT_SURNAME1);

            Assert.IsFalse(citizen1.CanSave());
            Assert.AreEqual(TestData.TEXT_NAME1, citizen1.Name);
            Assert.AreEqual(TestData.TEXT_SURNAME1, citizen1.Surname);

            WinRTCompV2.IAddress address = citizen1.Address;
            address.City = TestData.TEXT_CTY1;
            address.Street = TestData.TEXT_STR1;
            address.ZipCode = TestData.TEXT_ZIP1;

            Assert.IsTrue(citizen1.CanSave());

            Assert.AreEqual(TestData.TEXT_CTY1, citizen1.Address.City);
            Assert.AreEqual(TestData.TEXT_STR1, citizen1.Address.Street);
            Assert.AreEqual(TestData.TEXT_ZIP1, citizen1.Address.ZipCode);

            WinRTCompV2.Address address2 = new WinRTCompV2.Address();
            WinRTCompV2.ICitizen citizen2 = new WinRTCompV2.Citizen(TestData.TEXT_NAME2, TestData.TEXT_SURNAME2, address2);

            Assert.IsFalse(citizen2.CanSave());
            Assert.AreEqual(TestData.TEXT_NAME2, citizen2.Name);
            Assert.AreEqual(TestData.TEXT_SURNAME2, citizen2.Surname);

            address2.City = TestData.TEXT_CTY2;
            address2.Street = TestData.TEXT_STR2;
            address2.ZipCode = TestData.TEXT_ZIP2;

            Assert.AreEqual(TestData.TEXT_CTY2, citizen2.Address.City);
            Assert.AreEqual(TestData.TEXT_STR2, citizen2.Address.Street);
            Assert.AreEqual(TestData.TEXT_ZIP2, citizen2.Address.ZipCode);
        }
    }
}
